pkg/storage/stores/tsdb/single_file_index.go

```diff
@@ -301,5 +301,48 @@ func (i *TSDBIndex) Stats(ctx context.Context, userID string, from, through mode
 		}
 		return p.Err()
 	})
+}
+
+func (i *TSDBIndex) LabelVolume(ctx context.Context, userID string, from, through model.Time, acc LabelVolumeAccumulator, shard *index.ShardAnnotation, shouldIncludeChunk shouldIncludeChunk, matchers ...*labels.Matcher) error {
+	volumes := make(map[string]map[string]uint64)
+
+	err := i.forPostings(ctx, shard, from, through, matchers, func(p index.Postings) error {
+		var ls labels.Labels
+		var filterer chunk.Filterer
+		if i.chunkFilter != nil {
+			filterer = i.chunkFilter.ForRequest(ctx)
+		}
+
+		for p.Next() {
+			fp, stats, err := i.reader.ChunkStats(p.At(), int64(from), int64(through), &ls)
+			if err != nil {
+				return err
+			}
+
+			// skip series that belong to different shards
+			if shard != nil && !shard.Match(model.Fingerprint(fp)) {
+				continue
+			}
+
+			if filterer != nil && filterer.ShouldFilter(ls) {
+				continue
+			}
```

Comment:
I'm torn on if we should allow filtering for these types of queries. I like it for parity with stats, but I also think given the use cases we know of today it might be nice to ship the first version without it. I'm curious if that would be a worthwhile performance trade off?